contracts/IAOLinearVesting.sol

```diff
@@ -0,0 +1,364 @@
+        }
+        offeringToken = _offeringToken;
+        // Setup block variables
+        startBlock = _startBlock;
+        endBlock = _startBlock + _endBlockOffset;
+        vestingEndBlock = endBlock + _vestingBlockOffset;
+        // Setup amount variables
+        offeringAmount = _offeringAmount;
+        raisingAmount = _raisingAmount;
+        totalAmount = 0;
+        // Setup admin variable
+        adminAddress = _adminAddress;
+    }
+
+    modifier onlyAdmin() {
+        require(msg.sender == adminAddress, "caller is not admin");
+        _;
+    }
+
+    modifier onlyActiveIAO() {
+        require(
+            block.number >= startBlock && block.number < endBlock,
+            "not iao time"
+        );
+        _;
+    }
+
+    function setOfferingAmount(uint256 _offerAmount) public onlyAdmin {
+        require(block.number < startBlock, "cannot update during active iao");
+        offeringAmount = _offerAmount;
+    }
+
+    function setRaisingAmount(uint256 _raisingAmount) public onlyAdmin {
+        require(block.number < startBlock, "cannot update during active iao");
+        raisingAmount = _raisingAmount;
+    }
+
+    /// @notice Deposits native EVM tokens into the IAO contract as per the value sent
+    ///   in the transaction.
+    function depositNative() external payable onlyActiveIAO {
+        require(isNativeTokenStaking, 'stake token is not native EVM token');
+        require(msg.value > 0, 'value not > 0');
+        depositInternal(msg.value);
+    }
+
+    /// @dev Deposit ERC20 tokens with support for reflect tokens
+    function deposit(uint256 _amount) external onlyActiveIAO {
+        require(!isNativeTokenStaking, "stake token is native token, deposit through 'depositNative'");
+        require(_amount > 0, "_amount not > 0");
+        uint256 pre = getTotalStakeTokenBalance();
+        stakeToken.safeTransferFrom(
+            msg.sender,
+            address(this),
+            _amount
+        );
+        uint256 finalDepositAmount = getTotalStakeTokenBalance() - pre;
+        depositInternal(finalDepositAmount);
+    }
+
+    /// @notice To support ERC20 and native token deposits this function does not transfer
+    ///  any tokens in, but only updates the state. Make sure to transfer in the funds
+    ///  in a parent function
+    function depositInternal(uint256 _amount) internal {
+        if (userInfo[msg.sender].amount == 0) {
+            addressList.push(msg.sender);
+        }
+        userInfo[msg.sender].amount += _amount;
+        totalAmount += _amount;
+        totalDebt += _amount;
+
+        emit Deposit(msg.sender, _amount);
+    }
+
+    function harvest() external nonReentrant {
+        require(block.number > endBlock, "not harvest time");
+        UserInfo storage currentUserInfo = userInfo[msg.sender];
+        require(currentUserInfo.amount > 0, "have you participated?");
+        require(currentUserInfo.lastBlockHarvested < vestingEndBlock, "nothing left to harvest");
```

Comment:
Could a user harvesting right before the `vestingEndBlock` lose one block worth of value?

Should this be `<=` ?

Reply:
If they harvest one block before then they would still be able to harvest on the `vestingEndBlock` as `lastBlockHarvested` would be set to the current block. Then they are still able to withdraw one more time any time in the future.